// @ionic/utils-terminal/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Debug = require("debug");
const debug = Debug('ionic:utils-terminal');
/**
 * These environment variables work for: Travis CI, CircleCI, Gitlab CI,
 * AppVeyor, CodeShip, Jenkins, TeamCity, Bitbucket Pipelines, AWS CodeBuild
 */
exports.CI_ENVIRONMENT_VARIABLES = ['CI', 'BUILD_ID', 'BUILD_NUMBER', 'BITBUCKET_COMMIT', 'CODEBUILD_BUILD_ARN'];
exports.CI_ENVIRONMENT_VARIABLES_DETECTED = exports.CI_ENVIRONMENT_VARIABLES.filter(v => !!process.env[v]);
if (exports.CI_ENVIRONMENT_VARIABLES_DETECTED.length > 0) {
    debug(`Environment variables for CI detected: ${exports.CI_ENVIRONMENT_VARIABLES_DETECTED.join(', ')}`);
}
exports.TERMINAL_INFO = Object.freeze({
    tty: Boolean(process.stdin.isTTY && process.stdout.isTTY && process.stderr.isTTY),
    ci: exports.CI_ENVIRONMENT_VARIABLES_DETECTED.length > 0,
    windows: process.platform === 'win32' || (process.env.OSTYPE && /^(msys|cygwin)$/.test(process.env.OSTYPE) ||
        process.env.MSYSTEM && /^MINGW(32|64)$/.test(process.env.MSYSTEM) ||
        process.env.TERM === 'cygwin'),
});
